proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,447 @@
+- the program account is writable, otherwise throw `InvalidArgument`
+- the provided authority (instruction account at index 1) signed,
+otherwise throw `MissingRequiredSignature`
+- the authority stored in the program account is the one provided,
+otherwise throw `IncorrectAuthority`
+- the status stored in the program account is not `Invalid`,
+otherwise throw `InvalidArgument`
+- the status stored in the program account is not `Finalized`,
+otherwise throw `Immutable`
+
+### Execution / Invocation
+
+Invoking programs owned by loader-v4 checks in the following order that:
+
+- the owner of the program account is loader-v4
+- the program account is at least as long enough for the header
+- the status stored in the program account is `Deployed` or `Finalized`
+- the program account was not deployed within the current slot (delay
+visibility)
+- the executable file stored in the program account passes executable
+verification
+
+failing any of the above checks must throw `UnsupportedProgramId`.
+
+### Program Management Instructions
+
+#### Write
+
+- Instruction accounts:
+  - `[writable]` The program account to write to.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `0u32`
+  - `u32` Byte offset at which to write the given bytes
+  - `[u8]` Chunk of the programs executable file
+- Behavior:
+  - Charge 32 + chunk_length_in_bytes / cpi_bytes_per_unit CUs
+  - Check there are at least two instruction accounts,
+  otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account
+  - Check the status stored in the program account is `NeverBeenDeployed` or
+  `Retracted`,
+  otherwise throw `InvalidArgument`
+  - Check that the end offset (sum of offset and length of the chunk) does
+  not exceed the maximum (program account length minus the header size),
+  otherwise throw `AccountDataTooSmall`
+  - Copy the chunk into the program account at the offset shifted by the
+  header size
+
+#### Copy
+
+- Instruction accounts:
+  - `[writable]` The program account to copy to.
+  - `[signer]` The authority of the program.
+  - `[]` The account to copy from.
+- Instruction data:
+  - Enum variant `1u32`
+  - `u32` Byte offset at which to write
+  - `u32` Byte offset at which to read
+  - `u32` Length of the chunk to copy in bytes
+- Behavior:
+  - Charge 32 + chunk_length_in_bytes / cpi_bytes_per_unit CUs
+  - Check there are at least three instruction accounts,
+  otherwise throw `NotEnoughAccountKeys`
+  - Check that program account and source account do not alias,
+  otherwise throw `AccountBorrowFailed`
+  - Verify the program account
+  - Check the status stored in the program account is `NeverBeenDeployed` or
+  `Retracted`,
+  otherwise throw `InvalidArgument`
+  - Check that the source account is owned by loader v2, v3 or v4,
+  otherwise throw `InvalidArgument`
+  - and look-up the source header size:
+    - loader-v2: 0 bytes
+    - loader-v3 buffer: 37 bytes
+    - loader-v3 programdata: 45 bytes
```

Comment:
Can you also specify that a non-buffer / non-programdata loader-v3 owned account will be rejected with `InvalidArgument`? (same for uninitialized loader-v4 account)